### PR TITLE
Add IAM support

### PR DESCRIPTION
**Pull request by lpatino10**
This PR adds in support for IAM in the core of the SDK. With these changes, users have the option of authenticating with Watson services using IAM tokens, which they can manage themselves or let the new `IamTokenManager` class handle.

Shout out to @dpopp07 for leading the way for this in the Node SDK. Definitely leaned on your changes for my implementation 😉

### CODE REVIEW

File: core/src/main/java/com/ibm/watson/developer_cloud/service/security/IamTokenManager.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+   */
+  private boolean isAccessTokenExpired() {
+    if (tokenData.getExpiresIn() == null || tokenData.getExpiration() == null) {
+      return true;
+    }
+
+    Double fractionOfTimeToLive = 0.8;
+    Long timeToLive = tokenData.getExpiresIn();
+    Long expirationTime = tokenData.getExpiration();
+    Double refreshTime = expirationTime - (timeToLive * (1.0 - fractionOfTimeToLive));
```
Comment (mamoonraja) [inline core/src/main/java/com/ibm/watson/developer_cloud/service/security/IamTokenManager.java]: can you review this one more time before merging?

File: core/src/test/resources/valid_iam_token.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+  "access_token": "aaaa-1111",
+  "refresh_token": "99999999",
+  "token_type": "Bearer",
+  "expires_in": 3600,
+  "expiration": 999999999999999999
```
Comment (dpopp07) [inline core/src/test/resources/valid_iam_token.json]: If someone is still maintaining this thousands of years from now, I hope they remember to change this expiration time!

File: core/src/main/java/com/ibm/watson/developer_cloud/service/WatsonService.java [hunk 3]
```diff
@@ -97,6 +101,15 @@
    */
   public WatsonService(final String name) {
     this.name = name;
+    String iamApiKey = CredentialUtils.getIAMKey(name);
+    String iamUrl = CredentialUtils.getIAMUrl(name);
+    if (iamApiKey != null && iamUrl != null) {
```
Comment (dpopp07) [inline core/src/main/java/com/ibm/watson/developer_cloud/service/WatsonService.java]: Even if the IAM API key is being pulled from VCAP credentials, the IAM URL is still optional, right? I think the `tokenManager` should still be created if there is an IAM API key but no IAM URL

File: core/src/main/java/com/ibm/watson/developer_cloud/service/WatsonService.java [hunk 4]
```diff
@@ -97,6 +100,10 @@
    */
   public WatsonService(final String name) {
     this.name = name;
+    String iamApiKey = CredentialUtils.getIAMKey(name);
+    if (iamApiKey != null) {
+      tokenManager = new IamTokenManager(new IamOptions.Builder().apiKey(iamApiKey).build());
+    }
     apiKey = CredentialUtils.getAPIKey(name);
```
[the remark this answers is not in the document]
Comment (germanattanasio) [reply core/src/main/java/com/ibm/watson/developer_cloud/service/WatsonService.java]: We need to add the URL to the VCAP services.

File: core/src/main/java/com/ibm/watson/developer_cloud/service/WatsonService.java (hunk 3 above)
Comment (germanattanasio) [reply core/src/main/java/com/ibm/watson/developer_cloud/service/WatsonService.java]: true

File: core/src/main/java/com/ibm/watson/developer_cloud/service/WatsonService.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (dpopp07) [reply core/src/main/java/com/ibm/watson/developer_cloud/service/WatsonService.java]: My apologies for the confusion, I am doing this in Node. Sounds good 👍

File: README.md [hunk 5]
```diff
@@ -123,17 +126,91 @@ credentials; the library will get them for you by looking at the [`VCAP_SERVICES
...
+
+```java
+// after instantiation
+VisualRecognition service = new VisualRecognition("2016-05-20");
+service.setApiKey("<api_key>");
+```
+
+### Using IAM
+
+When authenticating with IAM, you have the option of passing in the IAM API key, the IAM URL, and an IAM access token. **Be aware that passing in an access token means that you're assuming responsibility for maintaining that token's lifecycle.** If you instead pass in an IAM API key, the SDK will manage it for you.
```
[the remark this answers is not in the document]
Comment (lpatino10) [reply README.md]: No that's helpful! I was struggling to think of how to convey what was optional or not. I like your idea 👍

File: core/src/main/java/com/ibm/watson/developer_cloud/service/WatsonService.java (hunk 3 above)
Comment (dpopp07) [reply core/src/main/java/com/ibm/watson/developer_cloud/service/WatsonService.java]: @lpatino10 will you address this? As far as I can see, that is all that's left to be done

Comment (lpatino10) [reply core/src/main/java/com/ibm/watson/developer_cloud/service/WatsonService.java]: Oops, I missed this one. Making the change now.

**Pull request merged**